// telegram-mtproto/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MTProto = exports.MtpTimeManager = exports.setLogger = exports.ApiManager = exports.bin = exports.CryptoWorker = undefined;

var _crypto = require('./crypto');

Object.defineProperty(exports, 'CryptoWorker', {
  enumerable: true,
  get: function () {
    return _crypto.CryptoWorker;
  }
});

var _bin = require('./bin');

Object.defineProperty(exports, 'bin', {
  enumerable: true,
  get: function () {
    return _bin.bin;
  }
});

var _index = require('./service/api-manager/index');

Object.defineProperty(exports, 'ApiManager', {
  enumerable: true,
  get: function () {
    return _index.ApiManager;
  }
});

var _log = require('./util/log');

Object.defineProperty(exports, 'setLogger', {
  enumerable: true,
  get: function () {
    return _log.setLogger;
  }
});

var _wrap = require('./service/main/wrap');

var _wrap2 = _interopRequireDefault(_wrap);

var _timeManager = require('./service/time-manager');

var MtpTimeManager = _interopRequireWildcard(_timeManager);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.MtpTimeManager = MtpTimeManager;
exports.MTProto = _wrap2.default;
exports.default = _wrap2.default;
